[New Device Capabilities] Add `capabilities` to the device API (#6433)
PennyLaneAI/pennylane: doc/development/plugins.rst

Preprocessing
-------------

The :meth:`~.devices.Device.preprocess` method has two main responsibilities:

1) Create a :class:`~.TransformProgram` capable of turning an arbitrary batch of :class:`~.QuantumScript`\ s into a new batch of tapes supported by the ``execute`` method.
2) Setup the :class:`~.ExecutionConfig` dataclass by filling in device options and making decisions about differentiation.

> These two tasks can be extracted into private methods or helper functions if that improves source

Reviewer: We should probably make a mention here that that preprocessing be created by manually constructing the `preprocess` method, providing a toml file, or creating a default version from the toml file before extending and customizing the default behavior.

Developers shouldn't learn this information after they've already learned all about setting up a custom preprocessing.

Author: Done

* :func:`pennylane.devices.preprocess.validate_adjoint_trainable_params`
* :func:`pennylane.devices.preprocess.no_sampling`

> PennyLane also provides a default implementation of :meth:`~pennylane.devices.Device.preprocess`
> which should be sufficient for most plugin devices. To take advantage of this default implementation,
> you need to define a configuration file for your device as described below.

Reviewer: I'm not sure if this is the right way to phrase this.  It only has a default implementation if a proper toml file is provided.  It doesn't provided a good default preprocessing without it.

```suggestion
Instead of writing a customized :meth:`~pennylane.devices.Device.preprocess` method, plugin developers an instead provide a toml file configuration, and a default preprocessing will be populated from that file.
```

Or something like that.

Author: How about now?

.. _device_capabilities:

Device Capabilities
-------------------

> Optionally, you can add a ``config`` class variable pointing to your configuration file. This file

Reviewer: ```suggestion
Optionally, you can add a ``config_filepath`` class variable pointing to your configuration file. This file
```

should be a `toml file <https://toml.io/en/>`_ that describes which gates and features are supported

> by your device, i.e., what your of :meth:`~pennylane.devices.Device.execute` method accepts.

Reviewer: ```suggestion
by your device, i.e., what the :meth:`~pennylane.devices.Device.execute` method accepts.
```

.. code-block:: python

    from os import path
    from pennylane.devices import Device

    class MyDevice(Device):
        """My Documentation."""

        config_filepath = path.join(path.dirname(__file__), "relative/path/to/config.toml")

> This configuration file will be loaded into another class variable :attr:`~pennylane.devices.Device.capabilities`

Reviewer: Lets add a mention that the file should be imported as package data in `setup.py` and show how to do so.

Author: Done
